src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp

```diff
@@ -0,0 +1,245 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#include "mlpack/methods/preprocess/preprocess_string_util.hpp"
+#include <mlpack/core/util/mlpack_main.hpp>
+#include <mlpack/core/util/cli.hpp>
+#include <mlpack/core/data/string_encoding.hpp>
+#include <mlpack/core/data/tokenizers/split_by_any_of.hpp>
+#include <mlpack/core/data/string_encoding_policies/dictionary_encoding_policy.hpp>
+#include <mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp>
+#include <mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp>
+
+PROGRAM_INFO("preprocess_string_encoding",
+    // Short description.
+    "A utility to encode string data. This utility can encode string using "
+    "DictionaryEncoding, BagOfWordsEncoding and TfIdfEncoding methods.",
+    // Long description.
+    "This utility takes a dataset and the dimension and arguments and "
+    "encodes string dataset according to arguments given."
+    "\n\n"
+    "The dataset may be given as the file name and the output may be saved as "
+    + PRINT_PARAM_STRING("actual_dataset") + " and " +
+    PRINT_PARAM_STRING("preprocess_dataset") + " ."
+    "\n\n"
+    " Following arguments may be given " + PRINT_PARAM_STRING("encoding_type") +
+    " to encode the dataset using a specific encoding type and " + " Also the "
+    "dimension which contains the string dataset " +
+    PRINT_PARAM_STRING("dimension") + "."
+    "\n\n"
+    "So, a simple example where we want to encode string dataset " +
+    PRINT_DATASET("X") + ", which is having string data in its 3 Column,"
+    " using DictionaryEncoding as encoding type."
+    "\n\n" +
+    PRINT_CALL("preprocess_string", "actual_dataset", "X",
+        "preprocess_dataset", "X", "dimension", 3, "encoding_type",
+        "DictionaryEncoding") +
+    "\n\n",
+    SEE_ALSO("@preprocess_binarize", "#preprocess_binarize"),
+    SEE_ALSO("@preprocess_describe", "#preprocess_describe"),
+    SEE_ALSO("@preprocess_imputer", "#preprocess_imputer"));
+
+PARAM_STRING_IN_REQ("actual_dataset", "File containing the reference dataset.",
+    "t");
+PARAM_STRING_IN_REQ("preprocess_dataset", "File containing the preprocess or"
+    "encoded dataset.", "o");
+PARAM_STRING_IN("column_delimiter", "delimeter used to seperate Column in files"
+    "example '\\t' for '.tsv' and ',' for '.csv'.", "d", "\t");
+PARAM_STRING_IN("delimiter", "A set of chars that is used as delimeter to"
+    "tokenize the string dataset ", "D", " ");
+PARAM_STRING_IN("encoding_type", "Type of encoding", "e", "DictionaryEncoding")
+PARAM_STRING_IN("tfidf_encoding_type", "Type of tfidf encoding", "E",
+    "RawCount")
+PARAM_FLAG("smooth_idf", "True to have smooth_idf for Tf-Idf.", "s");
+PARAM_VECTOR_IN_REQ(std::string, "dimension", "Column which contains the"
+    "string data. (1 by default)", "c");
+
```

Comment:
Yeah, unfortunately the bag of words algorithm, and tf-idf don't preserve ordering. However dictionary encoding does. I think we could add the `Decode()` function to the encoder, but we should point out that the encoding isn't strictly one-to-one.

Reply:
Can we come up with anything in this world to dodge that ??? 
Like any hack or something breakthrough 🤯

I will try to give it a thought

But we should adapt something similar to our scaling api


Scaler.fit()
Scaler.transform()
Scaler.detransform()

Because decide is necessary

Reply:
For BoW and tf-idf it's not reversible, and that's okay---the way those are typically used in practice is that the result is never mapped back to the original space (it can't be).  Decoding, I guess, would only make sense for dictionary encoding (or even one-hot dictionary encoding?).